design/ceph/object/ceph-cosi-driver.md

```diff
@@ -0,0 +1,150 @@
+
+The Ceph COSI driver is deployed as a statefulset with a single replica along with [COSI sidecar container](https://github.com/kubernetes-sigs/container-object-storage-interface-provisioner-sidecar). The Ceph COSI driver is deployed in the same namespace as the Rook operator. The Ceph COSI driver is deployed with a service account that has the following RBAC permissions:
+
+## Integration plan with Rook
+
+The aim to support the v1alpha1 version of COSI in Rook v1.12. It will be extended to beta and release versions as appropriate. There will be one COSI Driver support from Rook. The driver will be started automatically with default settings when first Object Store gets created. The driver will be stopped when last Object Store is deleted. The driver will be deployed in the same namespace as Rook operator. The user can provide additional setting via new `CephCOSIDriver` CRD which is owned by Rook.
+
+### Pre-requisites
+
+- COSI CRDs should be installed in the cluster via following command
+
+```bash
+kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-api
+```
+
+- COSI controller should be deployed in the cluster via following command
+
+```bash
+kubectl apply -k github.com/kubernetes-sigs/container-object-storage-interface-controller
+```
+
+### End to End Workflow
+
+#### Changes in Common.yaml
+
+Following contents need to be append to `common.yaml` :
+
+- <https://github.com/ceph/ceph-cosi/blob/master/resources/sa.yaml>
+- <https://github.com/ceph/ceph-cosi/blob/master/resources/rbac.yaml>
+
+#### ceph-object-cosi-controller
+
+The `ceph-object-cosi-controller` will span Ceph COSI Driver pod with default settings on the Rook Operator Namespace. The controller will bring up the Ceph COSI driver when first object store is created and will stop the COSI driver when last object store is deleted only if it detect COSI Controller is running on default namespace. The controller will also watch for CephCosiDriver CRD, if defined the driver will be started with the settings provided in the CRD. If the Ceph COSI driver if up and running,it  will also create `CephObjectStoreUser` named  `cosi` for each object store which internally creates a secret rook-ceph-object-user-<objectstore-name>-cosi provides credentials for the object store. This can be specified in the BucketClass and BucketAccessClass. Also this controller ensures maximum one CephCosiDriver CRD exists in the cluster. For v1.12 supports object store belong to RGW multisite are not supported.
+
+#### cephcosidriver.ceph.rook.io CRD
+
+The users can define following CRD so that configuration related the Ceph COSI driver can be passed. This is not mandatory and the driver will be started with default settings if this CRD is not defined.
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephCosiDriver
+metadata:
+  name: rook-ceph-cosi
+  namespace: rook-ceph
+spec:
```

Comment:
We need these in the spec:
- `placement`: for node affinity, tolerations, and so on similar to the CephCluster and other CRs
- `resources`: for memory/cpu